apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +83,211 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
+	message = truncateLongMessage(message)
+
+	detailedReason := warnType + reason
+	dc.updateReason(detailedReason, message)
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		dc.Warnings = append(dc.Warnings, SubCondition{
+			Type:    warnType,
+			Status:  ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		})
+		return
+	}
+
+	// If the warning is already present, just update the reason and message.
+	dc.Warnings[i].updateReason(reason, message)
+}
+
+// AddWarningf adds an warning-level Subcondition to the DetailedCondition, using
+// fmt.Sprintf on the formatmsg and args params.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddWarningf(warnType, reason, formatmsg string, args ...interface{}) {
+	dc.AddWarning(warnType, reason, fmt.Sprintf(formatmsg, args...))
+}
+
+// GetWarning gets an warning of the given warnType.
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetWarning(warnType string) (SubCondition, bool) {
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Warnings[i], true
+}
+
+// updateReason updates a DetailedCondition's reason and message correctly
+// if they are different to the existing ones.
+// Note that this helper may be used to build the first iteration of a
+// DetailedCondition, so it's possible that `Reason` and/or `Message` will
+// be empty when it is called.
+func (dc *DetailedCondition) updateReason(reason, message string) {
+	if dc.Reason == "" {
+		dc.Reason = reason
+		dc.Message = message
+		return
+	}
+
+	if dc.Reason != reason {
+		dc.Reason = "MultipleProblems"
+		dc.Message = "Multiple problems were found, see errors or warnings for details"
+		return
+	}
+
+	// This case covers the same `Reason` being used multiple times.
+	// The only case for this is if we're adding more details about multiple errors for the same reason.
+	if dc.Message != message {
+		// Only add the message if it's not already in there somewhere.
+		if !strings.Contains(dc.Message, message) {
+			dc.Message = dc.Message + ", " + message
+		}
+	}
+}
+
+// IsPositivePolarity returns true if the DetailedCondition is a positive-polarity
+// condition like `Valid` or `Ready`, and false otherwise.
+func (dc *DetailedCondition) IsPositivePolarity() bool {
+	switch dc.Type {
+	case ValidConditionType:
+		return true
+	default:
+		return false
+	}
+}
+
+func (sc *SubCondition) updateReason(reason, message string) {
+	if sc.Reason == "" {
+		sc.Reason = reason
+		sc.Message = message
+		return
+	}
+
+	if sc.Reason != reason {
+		sc.Reason = "MultipleReasons"
+	}
+
+	if !strings.Contains(sc.Message, message) {
+		sc.Message = sc.Message + ", " + message
```

Comment:
This can blow the validation max length.

Reply:
If we keep this, I'll put a truncate afterwards then.